components/places/src/bookmark_sync/store.rs

```diff
@@ -0,0 +1,1723 @@
+        self.store.stage_local_items_to_upload()?;
+        self.store.db.execute_batch(
+            "
+            DELETE FROM mergedTree;
+            DELETE FROM idsToWeaklyUpload;",
+        )?;
+        tx.commit()?;
+        Ok(())
+    }
+}
+
+/// A helper that interpolates a named SQL common table expression (CTE) for
+/// local items. The CTE may be included in a `WITH RECURSIVE` clause.
+struct LocalItemsFragment<'a>(&'a str);
+
+impl<'a> fmt::Display for LocalItemsFragment<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "{name}(id, guid, parentId, parentGuid, position, type, title, parentTitle,
+                    placeId, dateAdded, lastModified, syncChangeCounter, level) AS (
+             SELECT b.id, b.guid, 0, NULL, b.position, b.type, b.title, NULL,
+                    b.fk, b.dateAdded, b.lastModified, b.syncChangeCounter, 0
+             FROM moz_bookmarks b
+             WHERE b.guid = '{root_guid}'
+             UNION ALL
+             SELECT b.id, b.guid, s.id, s.guid, b.position, b.type, b.title, s.title,
+                    b.fk, b.dateAdded, b.lastModified, b.syncChangeCounter, s.level + 1
+             FROM moz_bookmarks b
+             JOIN {name} s ON s.id = b.parent)",
+            name = self.0,
+            root_guid = BookmarkRootGuid::Root.as_guid().as_ref()
+        )
+    }
+}
+
+fn type_to_kind<'a>(typ: &'a str, url: UrlOrPlaceId<'a>) -> TypeToKind<'a> {
+    TypeToKind { typ, url }
+}
+
+/// A helper that interpolates a SQL expression for converting Places item types
+/// to Sync record kinds. `typ` is the name of the bookmark type column in the
+/// projection.
+struct TypeToKind<'a> {
+    typ: &'a str,
+    url: UrlOrPlaceId<'a>,
+}
+
+impl<'a> fmt::Display for TypeToKind<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            r#"(CASE {typ}
+                WHEN {bookmark_type} THEN (
+                    CASE substr({url}, 1, 6)
+                    /* Queries are bookmarks with a "place:" URL scheme. */
+                    WHEN 'place:' THEN {query_kind}
+                    ELSE {bookmark_kind}
+                    END
+                )
+                WHEN {folder_type} THEN {folder_kind}
+                ELSE {separator_kind}
+                END)"#,
+            typ = self.typ,
+            bookmark_type = BookmarkType::Bookmark as u8,
+            url = self.url,
+            bookmark_kind = SyncedBookmarkKind::Bookmark as u8,
+            folder_type = BookmarkType::Folder as u8,
+            folder_kind = SyncedBookmarkKind::Folder as u8,
+            separator_kind = SyncedBookmarkKind::Separator as u8,
+            query_kind = SyncedBookmarkKind::Query as u8
+        )
+    }
+}
+
+/// A helper that interpolates a SQL expression for a Place URL. This avoids a
+/// subquery if the URL is already available in the projection.
+enum UrlOrPlaceId<'a> {
```

Comment:
I think it would be easy for someone to assume this is supposed to represent either a Url or a Guid (or something), rather than what they are, which are column names. I can think of three ways to help avoid this:

1. Note in the comment that these are suppsed to be column references and not e.g. urls, or place IDs
2. Force these to be &'static str instead of parameterizing on lifetime. We should always be providing strings known at compile time (if tests need an escape hatch they can always go through a box and Box::leak the string to get a static str or something, but I'd be surprised if it was needed)
3. Maybe suffix types like this with `Expr` or `Fragment` to indicate they're effectively SQL expressions/fragments?

Reply:
did *both* 'static and *Fragment!